flink-tests/src/test/java/org/apache/flink/test/recovery/TaskManagerDisconnectOnShutdownITCase.java

```diff
@@ -0,0 +1,304 @@
+import org.apache.flink.runtime.resourcemanager.ArbitraryWorkerResourceSpecFactory;
+import org.apache.flink.runtime.resourcemanager.ResourceManager;
+import org.apache.flink.runtime.resourcemanager.ResourceManagerFactory;
+import org.apache.flink.runtime.resourcemanager.ResourceManagerRuntimeServices;
+import org.apache.flink.runtime.resourcemanager.ResourceManagerRuntimeServicesConfiguration;
+import org.apache.flink.runtime.resourcemanager.StandaloneResourceManager;
+import org.apache.flink.runtime.resourcemanager.StandaloneResourceManagerFactory;
+import org.apache.flink.runtime.rpc.FatalErrorHandler;
+import org.apache.flink.runtime.rpc.RpcService;
+import org.apache.flink.runtime.taskexecutor.SlotReport;
+import org.apache.flink.runtime.taskexecutor.TaskExecutorResourceUtils;
+import org.apache.flink.runtime.taskexecutor.TaskManagerRunner;
+import org.apache.flink.test.util.TestProcessBuilder;
+import org.apache.flink.util.ConfigurationException;
+import org.apache.flink.util.TestLogger;
+
+import org.junit.Test;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+
+import java.time.Duration;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static org.apache.flink.runtime.testutils.CommonTestUtils.getJavaCommandPath;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.fail;
+
+/** This test ensures the TaskManager disconnects from the ResourceManager on shutdown. */
+public class TaskManagerDisconnectOnShutdownITCase extends TestLogger {
+
+    @Test
+    public void testTaskManagerProcessFailure() throws Exception {
+        Configuration config = new Configuration();
+        config.setString(JobManagerOptions.ADDRESS, "localhost");
+
+        // disable heartbeats
+        config.set(HeartbeatManagerOptions.HEARTBEAT_RPC_FAILURE_THRESHOLD, -1);
+        config.setInteger(TaskManagerOptions.NUM_TASK_SLOTS, 2);
+
+        config.set(TaskManagerOptions.MANAGED_MEMORY_SIZE, MemorySize.parse("4m"));
+        config.set(TaskManagerOptions.NETWORK_MEMORY_MIN, MemorySize.parse("3200k"));
+        config.set(TaskManagerOptions.NETWORK_MEMORY_MAX, MemorySize.parse("3200k"));
+        config.set(TaskManagerOptions.TASK_HEAP_MEMORY, MemorySize.parse("128m"));
+        config.set(TaskManagerOptions.CPU_CORES, 1.0);
+        config.setString(JobManagerOptions.EXECUTION_FAILOVER_STRATEGY, "full");
+        config.set(JobManagerOptions.RESOURCE_WAIT_TIMEOUT, Duration.ofSeconds(30L));
+
+        // check that we run this test only if the java command
+        // is available on this machine
+        String javaCommand = getJavaCommandPath();
+        if (javaCommand == null) {
+            fail("cannot find java executable");
+        }
+
+        TaskManagerConnectionTracker tracker = new TaskManagerConnectionTracker();
```

Comment:
nit: just a good practice to follow, always mark all the variables as `final` unless you intend to mutate them. It's bit more verbose, but it gives a good hint to anyone reading the code and it's less error prone in general.

Reply:
true